docs/docs/typelevel.md

```diff
@@ -0,0 +1,532 @@
+`ToNat(n - 1)`. This gives `S[Nat]` as the type of the else
+clause. The full type of the right hand side `Z | S[Nat]` would then
+be checked against the upper approximation of the declared result type
+`ToNat(n)`, which is again `Nat`. This succeeds, so the definition is deemed type-correct.
+
+
+### Second Step: Nicer Syntax for Type Definitions
+
+The syntax for `ToNat` was a bit clunky. This is no outlier - it turns out
+that most inline type definitions use similar staging constructs. All
+parameters tend to be inline, and the body is typically a conditional
+expression which returns a quoted type in each branch. We can cut down
+this boilerplate by introducing a shorthand form for parameterized
+type definitions, which allows us to express `ToNat` as follows:
+
+    type ToNat(n: Int) <: Nat =
+      if n == 0 then Z
+      else S[ToNat(n - 1)]
+
+The short form expands precisely to the previous definition of `ToNat`.
+To get from a short form type definition to a long form type macro,
+the following rewritings are performed:
+
+ - `type` is expanded to `inline type def`.
+ - All value parameters get an `inline` modifier.
+ - If the right hand side consists of a conditional, it is put in a spliced block `~{...}`.
+ - The right hand sides of all conditionals are put in type quotes `'[...]`.
+
+We expect that most computed types coming up in practice will use the
+short form. The long form is kept around, to cater for more advanced
+use cases, and to give a precise meaning to the short form.
+
+### Third Step: Query Conditionals
+
+What about defining the inverse of `ToNat`? This should be a function
+that takes a _type_ of peano numbers and produces a compile-time
+integer _value_. We can emulate the effect of such a function with
+implicits but it's a bit roundabout since we would need several
+overloaded methods each taking a different implicit parameter. A more
+direct, closed formulation is possible if we allow implicit searches
+as conditions in spliced code. Example:
+
+    inline def toInt[N <: Nat]: Int = ~{
+      case N =:= Z => '0
+      case N =:= S[type N1] => '(1 + toInt[N1])
+    }
+
+A query conditional is written ` { <cases> }` where each case is of the form
+
+    case <type-pattern> <guard> => <expr>
+
+`<guard>` is either empty or of the form `if <expr>`. Some examples of legal cases are
+
+    case Ord[X] => e1
+    case X <:< List[Y] if n == 0 => e2
+    case _ if n < 0 => e3
+
+Query conditionals may only appear in spliced code. That sets them
+apart from blocks with normal case clauses which define partial
+functions and therefore can only appear in normal code.  Every query
+conditional is evaluated by the macro interpreter. The macro
+interpreter evaluates each case of a query conditional by performing
+an implicit search for the given type pattern. If the search succeeds,
+it continues with evaluating the guard, if there is one, followed by
+the right-hand side, otherwise it continues with the next case.
```

Comment:
This sounds like a mixture of logic and functional programming. How does this interact with backtracking in implicit search? After some experiences with this mixture I’d like a good answer, because this sort of combination can be pretty tricky to program (or too hard to use) with if you don’t have control of backtracking.

Reply:
The implicit query must be resolved at this point. If it is ambiguous it's an error. There is no backtracking over query conditionals.